File: Difficulty: Medium/Sort a stack/sort-a-stack.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

class SortedStack{
public:
	stack<int> s;
	void sort();
};

void printStack(stack<int> s)
{
    while (!s.empty())
    {
        printf("%d ", s.top());
       	s.pop();
    }
    printf("\n");
}

int main()
{
int t;
cin>>t;
while(t--)
{
	SortedStack *ss = new SortedStack();
	int n;
	cin>>n;
	for(int i=0;i<n;i++)
	{
	int k;
	cin>>k;
	ss->s.push(k);
	}
	ss->sort();
	printStack(ss->s);

cout << "~" << "\n";
}
}
// } Driver Code Ends


/*The structure of the class is
class SortedStack{
public:
    stack<int> s;
    void sort();
};
*/

/* The below method sorts the stack s
you are required to complete the below method */
void intsertelement(stack<int>& s, int element){
    if(s.empty() || element > s.top()){
        s.push(element);
        return;
    }
    
    // bahar nikalo or phir stack m dalo
    int temp = s.top();
    s.pop();
    intsertelement(s, element);
    s.push(temp);
    
}
void SortedStack ::sort() {
    // Your code here
    if(s.empty()) return;
    
    int top = s.top();
    s.pop();
    
    sort();
    
    intsertelement(s, top);
    
}

File: Difficulty: Medium/Sort a stack/sort-a-stack.java
class GfG {
    public void insertElement(Stack<Integer> s, int element) {
        if (s.isEmpty() || element > s.peek()) {
            s.push(element);
            return;
        }

        int temp = s.peek();
        s.pop();
        insertElement(s, element);
        s.push(temp);   
    }

    public Stack<Integer> sort(Stack<Integer> s) {
        if (s.isEmpty()) return s;

        int top = s.peek();
        s.pop();

        sort(s);

        insertElement(s, top);
        return s;  
    }
}
